### ISSUE TITLE
MQTT SSL connection not working on Windows

### ISSUE THREAD

**Issue by tinkerborg** [tinkerborg/thinq2-python#4]
Reported by @stefxx:

```Traceback (most recent call last):
File "example.py", line 79, in
thinq.mqtt.connect()
File "thinq2-python\thinq2\controller\mqtt.py", line 25, in connect
self.client.connect(endpoint.hostname, endpoint.port)
File "thinq2-python\thinq2\util_init_.py", line 10, in inner
memo[key] = func(*args, **kwargs)
File "thinq2-python\thinq2\controller\mqtt.py", line 44, in client
client.tls_set_context(self.ssl_context)
File "thinq2-python\thinq2\controller\mqtt.py", line 67, in ssl_context
context.load_cert_chain(certfile=client_cert.name, keyfile=private_key.name)
PermissionError: [Errno 13] Permission denied```

**Comment by tinkerborg** [comment]
Confirmed this is a cross platform issue. Reproduced this on a fresh Windows 10 VM, looking into it.

**Comment by tinkerborg** [comment]
TL;DR - This should work now! I like going into excess detail as a kindness to future folks googling similar issues. 

Excess detail: 

AWS IoT connections are basically MQTT over SSL with client cert auth. There is an LG API endpoint that accepts a CSR and signs a client cert, which you can then use to open an encrypted socket for MQTT. 

Annoyingly, creating an SSL context requires the certificate to exist in a file, because this is what the underlying OpenSSL API requires. I wanted the cert storage to be part of the application state and thus up to the user of this library. So, the library writes out the certificate, private key and CA cert to temporary files (CA cert could actually be loaded from a string directly, oddly enough...).

I used NamedTemporaryFile for this, as OpenSSL needs a path rather than a file handle. I then left the file handle open, the file will be removed when file handle is closed. Unfortunately, per [the docs](https://docs.python.org/3/library/tempfile.html), this prevents the file from being opened again (by OpenSSL) on Windows:

> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later). 

I've switched this code to use TemporaryDirectory instead, which creates a directory that will be recursively removed when the TemporaryDirectory is garbage collected. mkstemp() will securely create temp files, but not remove them when the fd is closed, so I used that to write cert/key files in the temp dir to initialize the context.

**Comment by tinkerborg** [comment]
@stefxx - assuming the above works correctly now, you should see events from your a/c unit. I'm assuming you don't also see events from v1 devices, can you confirm?

**Comment by stefxx** [comment]
You are great! Thank you so much!

b'{"data":{"state":{"reported":{"airState.tempState.current":23,"meta":{"allDeviceInfoUpdate":false,"messageId":"bvnxgMlMTT-rujeERl7qXQ"},"online":true,"static":{"countryCode":"NL","deviceType":"401"},"timestamp":1589004569281}}},"deviceId":"69xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx4a","type":"monitoring"}'

No, I don't see v1 events. I am no expert on MQTT, but looking at all the differences between v1 and v2 I don't expect MQTT to work for v1 devices. But I assume you need to subscribe each device to MQTT before you will receive any events? Maybe you can try to subscribe v1 devices as well?

And I like your "excess detail'!  Being a hobby programmer myself (VB.net is all I can do, and I really have a hard time understanding Python) it helps me understand it a lot better!

Thanks again. Looking forward to updates in the future. Especially device control ;)

**Comment by tinkerborg** [comment]
Didn't expect to see v1 events, but I was just curious. You don't need to subscribe to devices individually. When you hit the certificate endpoint, the response contains both the certificate PEM data and a list of subscriptions (which currently is just one). 

Yup, device control is coming. Next goal is to create an observable device controller abstraction that will retrieve a device, update its state based on MQTT events, and notify observers when the state changes. Then the next goal after that will be sending commands to the device.
